stake-pool/single-pool/src/processor.rs

```diff
@@ -0,0 +1,1078 @@
+        vote_account_address: &Pubkey,
+        user_stake_authority: &Pubkey,
+        token_amount: u64,
+    ) -> ProgramResult {
+        let account_info_iter = &mut accounts.iter();
+        let pool_stake_info = next_account_info(account_info_iter)?;
+        let pool_authority_info = next_account_info(account_info_iter)?;
+        let pool_mint_info = next_account_info(account_info_iter)?;
+        let user_stake_info = next_account_info(account_info_iter)?;
+        let user_token_account_info = next_account_info(account_info_iter)?;
+        let clock_info = next_account_info(account_info_iter)?;
+        let clock = &Clock::from_account_info(clock_info)?;
+        let token_program_info = next_account_info(account_info_iter)?;
+        let stake_program_info = next_account_info(account_info_iter)?;
+
+        check_pool_stake_address(program_id, vote_account_address, pool_stake_info.key)?;
+        let bump_seed = check_pool_authority_address(
+            program_id,
+            vote_account_address,
+            pool_authority_info.key,
+        )?;
+        check_pool_mint_address(program_id, vote_account_address, pool_mint_info.key)?;
+        check_token_program(token_program_info.key)?;
+        check_stake_program(stake_program_info.key)?;
+
+        // we enforce that our stake is active so depositing/withdrawing during the warmup epoch is impossible
+        let pre_pool_stake = get_active_stake_state(pool_stake_info, clock.epoch)?
+            .delegation
+            .stake;
+        msg!("Stake pre split {}", pre_pool_stake);
+
+        // we add initial lamports to supply to make the math work without minting tokens to incinerator
+        let token_supply = {
+            let pool_mint_data = pool_mint_info.try_borrow_data()?;
+            let pool_mint = Mint::unpack_from_slice(&pool_mint_data)?;
+            pool_mint.supply
+        };
+
+        // withdraw amount is determined off stake just like deposit amount
+        let withdraw_stake = calculate_withdraw_amount(token_supply, pre_pool_stake, token_amount)
+            .ok_or(SinglePoolError::UnexpectedMathError)?;
+
+        if withdraw_stake == 0 {
+            return Err(SinglePoolError::WithdrawalTooSmall.into());
+        }
+
+        // burn user tokens corresponding to the amount of stake they wish to withdraw
+        Self::token_burn(
+            vote_account_address,
+            token_program_info.clone(),
+            user_token_account_info.clone(),
+            pool_mint_info.clone(),
+            pool_authority_info.clone(),
+            bump_seed,
+            token_amount,
+        )?;
+
+        // split stake into a blank stake account the user has created for this purpose
+        Self::stake_split(
+            vote_account_address,
+            pool_stake_info.clone(),
+            pool_authority_info.clone(),
+            bump_seed,
+            withdraw_stake,
+            user_stake_info.clone(),
+        )?;
```

Comment:
This shouldn't happen, but as a sanity check, can you check that `withdraw_stake != pool_stake_info.lamports()`? Otherwise it's possible to take control of the pool's stake account by deleting it and resurrecting it in the same transaction